x-pack/plugins/lens/public/app_plugin/mounter.tsx

```diff
@@ -270,52 +309,161 @@ export async function mountApp(
   };
 }
 
-export function loadDocument(
+const getInitialDatasourceId = (datasourceMap: Record<string, Datasource>, doc?: Document) => {
+  return (doc && getActiveDatasourceIdFromDoc(doc)) || Object.keys(datasourceMap)[0] || null;
+};
+
+export function loadInitialStore(
   redirectCallback: (savedObjectId?: string) => void,
   initialInput: LensEmbeddableInput | undefined,
   lensServices: LensAppServices,
   lensStore: LensRootStore,
   embeddableEditorIncomingState: EmbeddableEditorState | undefined,
-  dashboardFeatureFlag: DashboardFeatureFlagConfig
+  dashboardFeatureFlag: DashboardFeatureFlagConfig,
+  datasourceMap: Record<string, Datasource>,
+  visualizationMap: Record<string, Visualization>,
+  initialContext?: VisualizeFieldContext
 ) {
   const { attributeService, chrome, notifications, data } = lensServices;
-  const { persistedDoc } = lensStore.getState().app;
+  const { persistedDoc, datasourceStates } = lensStore.getState().app;
   if (
     !initialInput ||
     (attributeService.inputIsRefType(initialInput) &&
       initialInput.savedObjectId === persistedDoc?.savedObjectId)
   ) {
+    initializeDatasources(datasourceMap, datasourceStates, undefined, initialContext, {
+      isFullEditor: true,
+    })
+      .then((result) => {
+        lensStore.dispatch(
+          setState({
+            datasourceStates: Object.entries(result).reduce(
+              (state, [datasourceId, datasourceState]) => ({
+                ...state,
+                [datasourceId]: {
+                  ...datasourceState,
+                  isLoading: false,
+                },
+              }),
+              {}
+            ),
+            isAppLoading: false,
+          })
+        );
+
+        {
```

Comment:
nitpick: why this isolated scope?

Reply:
@dej611 this code is not ready to be reviewed yet -> I marked it this way to remember about refactoring it. I'll let you know on Monday (hopefully) when it's ready!